Multisig/Features/Security/SensitiveEncryptedStore.swift

```diff
@@ -0,0 +1,183 @@
+//
+// Created by Dirk Jäckel on 15.11.22.
+// Copyright (c) 2022 Gnosis Ltd. All rights reserved.
+//
+
+import Foundation
+import CommonCrypto
+import CryptoKit
+
+class SensitiveEncryptedStore: EncryptedStore {
+
+    let keychainStorage: KeychainStorage
+
+    init(_ keychainStorage: KeychainStorage) {
+        self.keychainStorage = keychainStorage
+    }
+
+    init() {
+        keychainStorage = KeychainStorage()
+    }
+
+    // This is called, when using the new key security is activated after updating the app. Even if the user did not activate it.
+    func initialSetup() throws {
+        let derivedPasscode = try persistRandomPassword()
+        let sensitiveKey = try keychainStorage.createKeyPair()
+        try persistSensitivePublicKey(sensitiveKey: sensitiveKey)
+        try persistSensitivePrivateKey(derivedPasscode: derivedPasscode, sensitiveKey: sensitiveKey)
+    }
+
+    private func persistRandomPassword() throws -> String {
+        let randomPasscode = createRandomPassword()
+        let derivedPasscode = derivePasscode(from: randomPasscode)
+        try keychainStorage.storePasscode(derivedPasscode: derivedPasscode) // TODO check error?
+        return derivedPasscode
+    }
+
+    private func createRandomPassword() -> String {
+        Data.randomBytes(length: 32)!.toHexString()
+    }
+
+    private func persistSensitivePrivateKey(derivedPasscode: String, sensitiveKey: SecKey) throws { // create SE key (KEK) with a hard coded tag for example: "sensitive_KEK"
+        let sensitiveKEK = try keychainStorage.createSecureEnclaveKey(
+                useBiometry: false,
+                canChangeBiometry: true,
+                applicationPassword: derivedPasscode
+        )
+
+        // Convert SecKey -> Data SecKeyCopyExternalRepresentation -> CFData -> Data
+        // Copy private part of sensitive key
+        var error: Unmanaged<CFError>?
+        guard let sensitiveKeyData = SecKeyCopyExternalRepresentation(sensitiveKey, &error) else {
```

Comment:
We only use the public KEK to encrypt the sensitive key private key. That only happens when re-creating the KEK (password change, access change, etc) immediately **after** creating it. If we store this public KEK we will only retrieve it, when it is outdated. Because we just created a new KEK.

Reply:
OK, thank you for explanation, that is an important piece of information.